crates/accelerate/src/synthesis/linear/mod.rs

```diff
@@ -0,0 +1,192 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::QiskitError;
+use numpy::{IntoPyArray, PyArray2, PyReadonlyArray2, PyReadwriteArray2};
+use pyo3::prelude::*;
+
+mod utils;
+
+#[pyfunction]
+#[pyo3(signature = (mat, ncols=None, full_elim=false))]
+/// Gauss elimination of a matrix mat with m rows and n columns.
+/// If full_elim = True, it allows full elimination of mat[:, 0 : ncols]
+/// Returns the matrix mat, and the permutation perm that was done on the rows during the process.
+/// perm[0 : rank] represents the indices of linearly independent rows in the original matrix.
+/// Args:
+///     mat: a boolean matrix with n rows and m columns
+///     ncols: the number of columns for the gaussian elimination,
+///            if ncols=None, then the elimination is done over all the colums
+///     full_elim: whether to do a full elimination, or partial (upper triangular form)
+/// Returns:
+///     mat: the matrix after the gaussian elimination
+///     perm: the permutation perm that was done on the rows during the process
+fn gauss_elimination_with_perm(
+    py: Python,
+    mut mat: PyReadwriteArray2<bool>,
+    ncols: Option<usize>,
+    full_elim: Option<bool>,
+) -> PyResult<PyObject> {
+    let matmut = mat.as_array_mut();
+    let perm = utils::gauss_elimination_with_perm_inner(matmut, ncols, full_elim);
+    Ok(perm.to_object(py))
+}
+
+#[pyfunction]
+#[pyo3(signature = (mat, ncols=None, full_elim=false))]
+/// Gauss elimination of a matrix mat with m rows and n columns.
+/// If full_elim = True, it allows full elimination of mat[:, 0 : ncols]
+/// Returns the updated matrix mat.
+/// Args:
+///     mat: a boolean matrix with n rows and m columns
+///     ncols: the number of columns for the gaussian elimination,
+///            if ncols=None, then the elimination is done over all the colums
+///     full_elim: whether to do a full elimination, or partial (upper triangular form)
+/// Returns:
+///     mat: the matrix after the gaussian elimination
```

Comment:
It seems like the matrix is modified in-place and not returned 🙂 
```suggestion
/// This function modifies the input matrix in-place.
/// Args:
///     mat: a boolean matrix with n rows and m columns
///     ncols: the number of columns for the gaussian elimination,
///            if ncols=None, then the elimination is done over all the colums
///     full_elim: whether to do a full elimination, or partial (upper triangular form)
```